### ISSUE TITLE
Support Tensor parallel or pipeline parallel out of the box

### ISSUE THREAD

**Issue by blefaudeux** [facebookresearch/xformers#70]
# 🚀 Feature
Support tensor // or model // as a built in feature, through Fairscale ? cc @min-xu-ai @anj-s @suchenzang @VitaliyLi @iyerr3 

## Motivation
This is typically extra work for users, but squarely in the model space. xFormers could invest in a generic engineering effort here

## Pitch
Ideally, either a factory option or a dedicated model wrapper, and make this mostly transparent to users

## Alternatives
Not doing that

## Additional context
Discussed with @suchenzang

**Comment by anj-s** [comment]
This is something we have on our roadmap. Is it a different effort from what we have planned? 
1. We need to land the tensor parallel code that is already in FairScale. 
2. We need to complete/land the Distributed pipeline parallelism code that is already in FairScale.

**Comment by min-xu-ai** [comment]
I think it is useful to separate and make things bit more concrete:

1. pipeline: we are talking about pipelining identical attention layers, right? The non-attention layers at the beginning and the end of the pipeline will cause trouble with balance of both memory and speed. Also, if the number of layers is not evenly divisible by # of GPUs, again, balancing is an issue. What's the plan for addressing that?
2. tensor/megatron: this seems like more promising that pipeline since it is within a single layer. But it would be the first xformer layer that spread across multiple GPUs, right? Will that be breaking user's assumption when they use this kernel? Will you support both DDP and FSDP? There are quite a few things to iron out. Maybe a design doc would help?

**Comment by suchenzang** [comment]
@min-xu-ai Pipelining would look very similar to what is present in Megatron: https://github.com/NVIDIA/Megatron-LM/blob/b31e1296354e979722627a6c4dedafe19b51fa97/megatron/schedules.py#L187. In terms of balancing, it's the same issue as FSDP (materializing one layer at a time vs materializing full model across multiple GPUs), but with a bit more flexibility since the allocation across GPUs doesn't necessarily have to be a multiple of number of layers. With all these parallelism techniques, you do have to design your architecture to be "balanced" (decomposable into uniform "blocks"), which is where factorization methods will ideally be the workaround to help achieve that as well.

Definitely need to think about how to design this well in a modular / clean way.  The reason tensor / pipeline parallelism would be a great use-case to tackle is because this is where coupling with training codebases usually happen (distributed metrics collection having to be threaded through to avoid d2h transfers, etc.), and this tends to also be where integration with "external libraries" hit a wall (limited API / insufficient customization).  If this can somehow be cleanly solved through xFormers, this would be a huge step in decomposing monolithic training codebases.

**Comment by min-xu-ai** [comment]
> materializing one layer at a time vs materializing full model across multiple GPUs

can you elaborate on this? When FSDP materializing 1 layer on a gpu, it is overlapped with compute of another layer. The only requirement is that compute time needs to be long enough to keep all gpus busy. One can vary batch size to make compute longer. All gpus are doing the same thing. 

With pipeline, models with different layers, even transformer has embedding and output vocabulary layers, causing different gpus have different loads in flops and memory. My experience has been that pipeline balance was much much harder. But I was using a vision model. Shruti used transformer with pipeline as far as I remember. Perhaps she can share some experience here.

**Comment by suchenzang** [comment]
I'm mainly focused on data movement wrt params (/memory pressure). FSDP consumes more memory than necessary since it leaves memory allocation and deallocation to PyTorch, rather than deploying custom kernels to handle this explicitly. Overlapping with compute isn't helping, since that implies keeping the parameters associated with that layer in memory as well. Pipeline parallelism avoids this movement / memory pressure by keeping the parameters "fixed" in memory, and only moving batches of data through the "pipeline" (whereas the FSDP "pipeline" consists of moving both data and params).

**Comment by min-xu-ai** [comment]
I see what you mean. If by data you mean the activation, then I think FSDP actually keeps data fixed in memory, not moving in and out like that in the pipeline case. 

I wasn’t actually worried about data moving between gpus for pipeline. That movement is also overlapped between micro batches (except during the filling and draining bubbles. ) I am more worried about the structure of a pipeline requires each stage has about same compute latency. Or otherwise, gpus will be idle.

**Comment by suchenzang** [comment]
Ah yeah tuning various params (architecture, mainly, but microbatching frequency can also change between layers, potentially) to keep the latency balanced is tricky.  This is part where sacrificing some compute (idleness) may be worth it for that much more memory / bigger batch that you can use per gradient step. In that world even if you are 30% less compute efficient, if your model learns 30-40% faster since each gradient step is more "stable" (given more capacity and larger batch), then your total compute time spent to train to convergence is still the same. Sure it would be great if we kept compute efficiency maximal, but that isn't the only way to train a bigger/better model with the same compute budget.

**Comment by ClashLuke** [comment]
I'm sorry for not having anything of value I could add, but this would be super exciting, and I'm very much looking forward to seeing this added.﻿</br>
Do you think it'd be reasonable to add Mesh TensorFlow's dimension-based splitting to xformers (or PyTorch in general)?
